example/contrib/socks-client/socks/handshake.hpp

```diff
@@ -0,0 +1,53 @@
+//
+// Copyright (c) 2018 jackarain (jack dot wgm at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+#ifndef SOCKS_HANDSHAKE_HPP
+#define SOCKS_HANDSHAKE_HPP
+
+#include <socks/config.hpp>
+#include <socks/error.hpp>
+#include <boost/asio/async_result.hpp>
+#include <string>
+
+namespace socks {
+
+/** Perform the SOCKS v4 handshake in the client role.
+*/
+template<
+    typename AsyncStream,
+    typename Handler
+>
+BOOST_BEAST_ASYNC_RESULT1(Handler)
+async_handshake_v4(
+    AsyncStream& stream,
+    const std::string& hostname,
```

Comment:
should use `string_view` instead of `std::string const&`

Reply:
Interesting. The implementation stores a string in any case. From a stylstic persepctive I would support an argument for "pass string by value, move into place". In reality it will make no difference but conceptually it can avoid copying a long string.

In any case, will change as advised.

Reply:
label: comment-2

Reply:
I'm still not convinced it needs to store any string at all

Reply:
@vinniefalco Strictly correct for socks 4, socks 5 does not submit the hostname until after authentication, so the impl would need to either cache the supplied hostname or pre-build the buffer for submission in state 5. Similarly, auth credentials are not submitted in socks5 until state 2 so some buffering required there.

Storing a string is minutely inefficient, the tradeoff being maintainability (and the ability to add better debugging/loging logic in future if desired).

In production code I'd store the string, as I know at some point I'm going to want to add some logging or diagnostics. In library code where we are shipping shrink-wrapped guaranteed-correct black boxes, the 'correct' approach is probably a matter of opinion.

I will defer to your decision. You have a lot more experience in the maintenance of public libraries.

Reply:
Hmm...didn't know that about SOCKSv5. But just to keep things spicy, if `std::string` uses the small buffer optimization then passing it as a buffer to I/O routines from a composed operation could result in undefined behavior if the string is moved. Up to you to decide what you want to do. I would say, go for whatever is expedient. It is more important we get this merged than to make it "perfect."